kcrypt/lock.go

```diff
@@ -11,96 +13,142 @@ import (
-// the partition is decrypted first and the uuid changed after encryption so
-// any stored information needs to be updated (by the caller).
-func luksify(label string, logger types.KairosLogger, argsCreate ...string) (string, error) {
-	var pass string
-
-	// Make sure ghw will see all partitions correctly.
-	// older versions don't have --type=all. Try the simpler version then.
-	out, err := utils.SH("udevadm trigger --type=all || udevadm trigger")
-	if err != nil {
-		return "", fmt.Errorf("udevadm trigger failed: %w, out: %s", err, out)
-	}
-	syscall.Sync()
-
-	part, b, err := findPartition(label)
-	if err != nil {
-		logger.Err(err).Msg("find partition")
+// luksifyWithPassphrase encrypts a partition with an explicit passphrase (no plugin involved).
+func luksifyWithPassphrase(label string, passphrase string, logger types.KairosLogger, argsCreate ...string) (string, error) {
+	logger.Logger.Info().Msg("Running udevadm settle")
+	if err := udevAdmSettle(&logger, udevTimeout); err != nil {
 		return "", err
 	}
 
-	pass, err = getPassword(b)
+	logger.Logger.Info().Str("label", label).Msg("Finding partition")
+	info, err := findPartitionByLabel(label)
 	if err != nil {
-		logger.Err(err).Msg("get password")
+		logger.Err(err).Msg("find partition")
 		return "", err
 	}
 
-	mapper := fmt.Sprintf("/dev/mapper/%s", b.Name)
-	device := fmt.Sprintf("/dev/%s", part)
+	mapper := info.MapperPath()
+	device := info.DevicePath
 
 	extraArgs := []string{"--uuid", uuid.NewV5(uuid.NamespaceURL, label).String()}
 	extraArgs = append(extraArgs, "--label", label)
 	extraArgs = append(extraArgs, argsCreate...)
 
-	if err := createLuks(device, pass, extraArgs...); err != nil {
+	logger.Logger.Info().Str("device", device).Msg("Checking if device is mounted")
+	if err := unmountIfMounted(device, logger); err != nil {
+		logger.Err(err).Msg("unmount device")
+		return "", err
+	}
+
+	logger.Logger.Info().Str("device", device).Msg("Creating LUKS container")
+	if err := createLuks(device, passphrase, extraArgs...); err != nil {
 		logger.Err(err).Msg("create luks")
 		return "", err
 	}
 
-	err = formatLuks(device, b.Name, mapper, label, pass, logger)
+	logger.Logger.Info().Str("device", device).Str("label", label).Msg("Formatting LUKS container")
+	err = formatLuks(device, info.Partition.Name, mapper, label, passphrase, logger)
 	if err != nil {
 		logger.Err(err).Msg("format luks")
 		return "", err
 	}
 
-	return fmt.Sprintf("%s:%s:%s", b.FilesystemLabel, b.Name, b.UUID), nil
+	logger.Logger.Info().Str("label", label).Msg("Partition encryption completed")
+	return fmt.Sprintf("%s:%s:%s", info.Partition.FilesystemLabel, info.Partition.Name, info.Partition.UUID), nil
+}
+
+// unmountIfMounted checks if a device is mounted and unmounts it if needed.
+// This is necessary because cryptsetup cannot format a mounted partition.
+func unmountIfMounted(device string, logger types.KairosLogger) error {
```

Comment:
we get https://github.com/kairos-io/kairos-sdk/blob/main/mounts/system.go#L31 but im not sure if that would work here without too many changes. But I think the state.Runtime would have both the updated partitions, if they are mounted AND the list of encrypted partition by label and by device so maybe it would be useful to plug that in here as to not grow the code unnecessarily?